// Copyright  (C)  2007  Francois Cauwe <francois at cauwe dot org>

// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.

#include <stdio.h>
#include <iostream>
#include <chrono>
#include <cstdlib>
#include <memory>

#include "std_msgs/msg/float64_multi_array.hpp"
#include "sensor_msgs/msg/joint_state.hpp"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/wait_for_message.hpp"

#include "kdl_robot.h"
#include "kdl_control.h"
#include "kdl_planner.h"
#include "kdl_parser/kdl_parser.hpp"

using namespace KDL;
using FloatArray = std_msgs::msg::Float64MultiArray;
using namespace std::chrono_literals;

class Iiwa_pub_sub : public rclcpp::Node
{
public:
    Iiwa_pub_sub()
        : Node("ros2_kdl_node"),
          node_handle_(std::shared_ptr<Iiwa_pub_sub>(this))
    {
        // declare cmd_interface parameter (position, velocity)
        declare_parameter("cmd_interface", "position");
        get_parameter("cmd_interface", cmd_interface_);
        RCLCPP_INFO(get_logger(), "Current cmd interface is: '%s'", cmd_interface_.c_str());

        if (!(cmd_interface_ == "position" || cmd_interface_ == "velocity" || cmd_interface_ == "effort"))
        {
            RCLCPP_ERROR(get_logger(), "Selected cmd interface is not valid! Use 'position', 'velocity' or 'effort' instead...");
            return;
        }

        // declare traj_type parameter (linear, circular)
        declare_parameter("traj_type", "linear");
        get_parameter("traj_type", traj_type_);
        RCLCPP_INFO(get_logger(), "Current trajectory type is: '%s'", traj_type_.c_str());
        if (!(traj_type_ == "linear" || traj_type_ == "circular"))
        {
            RCLCPP_INFO(get_logger(), "Selected traj type is not valid!");
            return;
        }

        // declare s_type parameter (trapezoidal, cubic)
        declare_parameter("s_type", "trapezoidal");
        get_parameter("s_type", s_type_);
        RCLCPP_INFO(get_logger(), "Current s type is: '%s'", s_type_.c_str());
        if (!(s_type_ == "trapezoidal" || s_type_ == "cubic"))
        {
            RCLCPP_INFO(get_logger(), "Selected s type is not valid!");
            return;
        }
        // Parameters read from parameters.yaml file
        declare_parameter("total_time", 1.5);
        declare_parameter("trajectory_len", 150);
        declare_parameter("Kp", 5.0);
        get_parameter("total_time", this->total_time);
        get_parameter("trajectory_len", this->trajectory_len);
        get_parameter("Kp", this->Kp);
        RCLCPP_INFO(get_logger(), "Current total time is: '%f'", this->total_time);
        RCLCPP_INFO(get_logger(), "Current trajectory length is: '%d'", this->trajectory_len);
        RCLCPP_INFO(get_logger(), "Current Kp is: '%f'", this->Kp);

        declare_parameter("ctrl", "velocity_ctrl");
        get_parameter("ctrl", this->ctrl_mode_);
        RCLCPP_INFO(get_logger(), "Control mode selected: '%s'", this->ctrl_mode_.c_str());

        declare_parameter("lambda", 1.0);
        get_parameter("lambda", this->lambda_);
        RCLCPP_INFO(get_logger(), "Lambda value for null-space control: '%f'", this->lambda_);

        iteration_ = 0;
        t_ = 0;
        joint_state_available_ = false;

        // retrieve robot_description param
        auto parameters_client = std::make_shared<rclcpp::SyncParametersClient>(node_handle_, "robot_state_publisher");
        while (!parameters_client->wait_for_service(1s))
        {
            if (!rclcpp::ok())
            {
                RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
                rclcpp::shutdown();
            }
            RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
        }
        auto parameter = parameters_client->get_parameters({"robot_description"});

        // create KDLrobot structure
        KDL::Tree robot_tree;
        if (!kdl_parser::treeFromString(parameter[0].value_to_string(), robot_tree))
        {
            std::cout << "Failed to retrieve robot_description param!";
        }
        robot_ = std::make_shared<KDLRobot>(robot_tree);

        // Create joint array
        unsigned int nj = robot_->getNrJnts();
        KDL::JntArray q_min(nj), q_max(nj);
        q_min.data << -2.96, -2.09, -2.96, -2.09, -2.96, -2.09, -2.96;
        q_max.data << 2.96, 2.09, 2.96, 2.09, 2.96, 2.09, 2.96;
        robot_->setJntLimits(q_min, q_max);
        joint_positions_.resize(nj);
        joint_velocities_.resize(nj);
        joint_positions_cmd_.resize(nj);
        joint_velocities_cmd_.resize(nj);
        joint_efforts_cmd_.resize(nj);
        joint_efforts_cmd_.data.setZero();

        // Subscriber to jnt states
        jointSubscriber_ = this->create_subscription<sensor_msgs::msg::JointState>(
            "/joint_states", 10, std::bind(&Iiwa_pub_sub::joint_state_subscriber, this, std::placeholders::_1));

        // Wait for the joint_state topic
        while (!joint_state_available_)
        {
            RCLCPP_INFO(this->get_logger(), "No data received yet! ...");
            rclcpp::spin_some(node_handle_);
        }

        // Update KDLrobot object
        robot_->update(toStdVector(joint_positions_.data), toStdVector(joint_velocities_.data));
        KDL::Frame f_T_ee = KDL::Frame::Identity();
        robot_->addEE(f_T_ee);
        robot_->update(toStdVector(joint_positions_.data), toStdVector(joint_velocities_.data));

        // Compute EE frame
        init_cart_pose_ = robot_->getEEFrame();

        // Compute IK
        KDL::JntArray q(nj);
        robot_->getInverseKinematics(init_cart_pose_, q);

        // Initialize controller
        controller_ = std::make_unique<KDLController>(*robot_);

        Eigen::Vector3d init_position(Eigen::Vector3d(init_cart_pose_.p.data) - Eigen::Vector3d(0, 0, 0.1));
        Eigen::Vector3d end_position = Eigen::Vector3d::Zero();

        declare_parameter("end_position_x", 0.0);
        declare_parameter("end_position_y", 0.0);
        declare_parameter("end_position_z", 0.0);

        get_parameter("end_position_x", end_position[0]);
        get_parameter("end_position_y", end_position[1]);
        get_parameter("end_position_z", end_position[2]);
        RCLCPP_INFO(get_logger(), "End position loaded from params: [%.3f, %.3f, %.3f]",
                    end_position[0], end_position[1], end_position[2]);

        // Plan trajectory
        double traj_duration = 0, acc_duration = 0, traj_radius = 0.15;
        declare_parameter("traj_duration", 1.5);
        declare_parameter("acc_duration", 0.5);
        get_parameter("traj_duration", traj_duration);
        get_parameter("acc_duration", acc_duration);
        RCLCPP_INFO(get_logger(), "Current trajectory duration is: '%f'", traj_duration);
        RCLCPP_INFO(get_logger(), "Current acceleration duration is: '%f'", acc_duration);

        // Retrieve the first trajectory point
        if (traj_type_ == "linear")
        {
            planner_ = KDLPlanner(traj_duration, acc_duration, init_position, end_position); // currently using trapezoidal velocity profile
            if (s_type_ == "trapezoidal")
            {
                p_ = planner_.linear_traj_trapezoidal(t_);
            }
            else if (s_type_ == "cubic")
            {
                p_ = planner_.linear_traj_cubic(t_);
            }
        }
        else if (traj_type_ == "circular")
        {
            planner_ = KDLPlanner(traj_duration, init_position, traj_radius, acc_duration);
            if (s_type_ == "trapezoidal")
            {
                p_ = planner_.circular_traj_trapezoidal(t_);
            }
            else if (s_type_ == "cubic")
            {
                p_ = planner_.circular_traj_cubic(t_);
            }
        }

        if (cmd_interface_ == "position")
        {
            // Create cmd publisher
            cmdPublisher_ = this->create_publisher<FloatArray>("/iiwa_arm_controller/commands", 10);
            timer_ = this->create_wall_timer(std::chrono::milliseconds(100),
                                             std::bind(&Iiwa_pub_sub::cmd_publisher, this));

            // Send joint position commands
            for (long int i = 0; i < joint_positions_.data.size(); ++i)
            {
                desired_commands_[i] = joint_positions_(i);
            }
        }
        else if (cmd_interface_ == "velocity")
        {
            // Create cmd publisher
            cmdPublisher_ = this->create_publisher<FloatArray>("/velocity_controller/commands", 10);
            timer_ = this->create_wall_timer(std::chrono::milliseconds(100),
                                             std::bind(&Iiwa_pub_sub::cmd_publisher, this));

            // Set joint velocity commands
            for (long int i = 0; i < joint_velocities_.data.size(); ++i)
            {
                desired_commands_[i] = joint_velocities_(i);
            }
        }
        else if (cmd_interface_ == "effort")
        {
            // Create cmd publisher
            cmdPublisher_ = this->create_publisher<FloatArray>("/effort_controller/commands", 10);
            timer_ = this->create_wall_timer(std::chrono::milliseconds(100),
                                             std::bind(&Iiwa_pub_sub::cmd_publisher, this));

            // Set joint effort commands
            for (long int i = 0; i < joint_efforts_cmd_.data.size(); ++i)
            {
                desired_commands_[i] = joint_efforts_cmd_(i);
            }
        }

        // Create msg and publish
        std_msgs::msg::Float64MultiArray cmd_msg;
        cmd_msg.data = desired_commands_;
        cmdPublisher_->publish(cmd_msg);

        RCLCPP_INFO(this->get_logger(), "Starting trajectory execution ...");
    }

private:
    // define trajectory
    double total_time = 6.0; //
    int trajectory_len = 0;  //
    double Kp = 0;

    void cmd_publisher()
    {
        int loop_rate = trajectory_len / total_time;
        double dt = 1.0 / loop_rate;
        t_ += dt;
        iteration_ = iteration_ + 1;

        if (t_ < total_time)
        {
            if (traj_type_ == "linear")
            {
                if (s_type_ == "trapezoidal")
                {
                    p_ = planner_.linear_traj_trapezoidal(t_);
                }
                else if (s_type_ == "cubic")
                {
                    p_ = planner_.linear_traj_cubic(t_);
                }
            }
            else if (traj_type_ == "circular")
            {
                if (s_type_ == "trapezoidal")
                {
                    p_ = planner_.circular_traj_trapezoidal(t_);
                }
                else if (s_type_ == "cubic")
                {
                    p_ = planner_.circular_traj_cubic(t_);
                }
            }

            // Compute EE frame
            KDL::Frame cartpos = robot_->getEEFrame();

            // Compute desired Frame
            KDL::Frame desFrame;
            desFrame.M = cartpos.M;
            desFrame.p = toKDL(p_.pos);

            // compute errors
            Eigen::Vector3d error = computeLinearError(p_.pos, Eigen::Vector3d(cartpos.p.data));
            Eigen::Vector3d o_error = computeOrientationError(toEigen(init_cart_pose_.M), toEigen(cartpos.M));
            std::cout << "The error norm is : " << error.norm() << std::endl;

            if (cmd_interface_ == "position")
            {
                // Next Frame
                KDL::Frame nextFrame;
                nextFrame.M = cartpos.M;
                nextFrame.p = cartpos.p + (toKDL(p_.vel) + toKDL(Kp * error)) * dt;

                // Compute IK
                joint_positions_cmd_ = joint_positions_;
                robot_->getInverseKinematics(nextFrame, joint_positions_cmd_);
            }
            else if (cmd_interface_ == "velocity")
            {
                if (this->ctrl_mode_ == "velocity_ctrl")
                {
                    Vector6d cartvel;
                    cartvel << p_.vel + Kp * error, o_error;
                    joint_velocities_cmd_.data = pseudoinverse(robot_->getEEJacobian().data) * cartvel;
                }
                else if (this->ctrl_mode_ == "velocity_ctrl_null")
                {
                    joint_velocities_cmd_.data = this->controller_->velocity_ctrl_null(desFrame, this->Kp, this->lambda_);
                }
            }
            else if (cmd_interface_ == "effort")
            {
                joint_efforts_cmd_.data[0] = 0.1 * std::sin(2 * M_PI * t_ / total_time);
            }

            // Update KDLrobot structure
            robot_->update(toStdVector(joint_positions_.data), toStdVector(joint_velocities_.data));

            if (cmd_interface_ == "position")
            {
                // Set joint position commands
                for (long int i = 0; i < joint_positions_.data.size(); ++i)
                {
                    desired_commands_[i] = joint_positions_cmd_(i);
                }
            }
            else if (cmd_interface_ == "velocity")
            {
                // Set joint velocity commands
                for (long int i = 0; i < joint_velocities_.data.size(); ++i)
                {
                    desired_commands_[i] = joint_velocities_cmd_(i);
                }
            }
            else if (cmd_interface_ == "effort")
            {
                // Set joint effort commands
                for (long int i = 0; i < joint_efforts_cmd_.data.size(); ++i)
                {
                    desired_commands_[i] = joint_efforts_cmd_(i);
                }
            }

            // Create msg and publish
            std_msgs::msg::Float64MultiArray cmd_msg;
            cmd_msg.data = desired_commands_;
            cmdPublisher_->publish(cmd_msg);
        }
        else
        {
            RCLCPP_INFO_ONCE(this->get_logger(), "Trajectory executed successfully ...");

            // Send joint velocity commands
            if (cmd_interface_ == "position")
            {
                // Set joint position commands
                for (long int i = 0; i < joint_positions_.data.size(); ++i)
                {
                    desired_commands_[i] = joint_positions_cmd_(i);
                }
            }
            else if (cmd_interface_ == "velocity")
            {
                // Set joint velocity commands
                for (long int i = 0; i < joint_velocities_.data.size(); ++i)
                {
                    desired_commands_[i] = 0.0;
                }
            }
            else if (cmd_interface_ == "effort")
            {
                // Set joint effort commands
                for (long int i = 0; i < joint_efforts_cmd_.data.size(); ++i)
                {
                    desired_commands_[i] = joint_efforts_cmd_(i);
                }
            }

            // Create msg and publish
            std_msgs::msg::Float64MultiArray cmd_msg;
            cmd_msg.data = desired_commands_;
            cmdPublisher_->publish(cmd_msg);
        }
    }

    void joint_state_subscriber(const sensor_msgs::msg::JointState &sensor_msg)
    {
        joint_state_available_ = true;
        for (unsigned int i = 0; i < sensor_msg.position.size(); i++)
        {
            joint_positions_.data[i] = sensor_msg.position[i];
            joint_velocities_.data[i] = sensor_msg.velocity[i];
        }
    }

    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr jointSubscriber_;
    rclcpp::Publisher<FloatArray>::SharedPtr cmdPublisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::TimerBase::SharedPtr subTimer_;
    rclcpp::Node::SharedPtr node_handle_;

    std::vector<double> desired_commands_ = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    KDL::JntArray joint_positions_;
    KDL::JntArray joint_velocities_;

    KDL::JntArray joint_positions_cmd_;
    KDL::JntArray joint_velocities_cmd_;
    KDL::JntArray joint_efforts_cmd_;

    std::shared_ptr<KDLRobot> robot_;
    KDLPlanner planner_;

    trajectory_point p_;

    int iteration_;
    bool joint_state_available_;
    double t_;
    std::string cmd_interface_;
    std::string traj_type_;
    std::string s_type_;

    KDL::Frame init_cart_pose_;

    std::string ctrl_mode_;
    double lambda_;
    std::unique_ptr<KDLController> controller_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Iiwa_pub_sub>());
    rclcpp::shutdown();
    return 1;
}